1. Open and look through the data file

In [ ]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score

df=pd.read_csv('/datasets/users_behavior.csv')
print(df.info())
print(df.head(20))


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3214 entries, 0 to 3213
Data columns (total 5 columns):
calls       3214 non-null float64
minutes     3214 non-null float64
messages    3214 non-null float64
mb_used     3214 non-null float64
is_ultra    3214 non-null int64
dtypes: float64(4), int64(1)
memory usage: 125.7 KB
None
    calls  minutes  messages   mb_used  is_ultra
0    40.0   311.90      83.0  19915.42         0
1    85.0   516.75      56.0  22696.96         0
2    77.0   467.66      86.0  21060.45         0
3   106.0   745.53      81.0   8437.39         1
4    66.0   418.74       1.0  14502.75         0
5    58.0   344.56      21.0  15823.37         0
6    57.0   431.64      20.0   3738.90         1
7    15.0   132.40       6.0  21911.60         0
8     7.0    43.39       3.0   2538.67         1
9    90.0   665.41      38.0  17358.61         0
10   82.0   560.51      20.0   9619.53         1
11   45.0   344.32      13.0  19898.81         0
12   51.0   437.13      61.0  21

2. Split the source data into a training set, a validation set, and a test set.

In [ ]:
#defining the features and target for train, valid and test data

train,df_test=train_test_split(df,test_size=0.2, random_state=12345)
df_train, df_valid=train_test_split(train, test_size=0.2, random_state=12345)

features_train=df_train.drop(['is_ultra'], axis=1)
target_train=df_train['is_ultra']

features_test=df_test.drop(['is_ultra'], axis=1)
target_test=df_test['is_ultra']

features_valid=df_valid.drop(['is_ultra'], axis=1)
target_valid=df_valid['is_ultra']

3. Investigate the quality of different models by changing hyperparameters.

In [ ]:
#hyperparameter tuning for the Decision Tree

for depth in range(1,6):
    model_dt=DecisionTreeClassifier(max_depth=depth,random_state=12345)
    model_dt.fit(features_train, target_train)
    score_dt=model_dt.score(features_valid, target_valid)
    print("max_depth:",depth,":",score_dt)

max_depth: 1 : 0.7223300970873786
max_depth: 2 : 0.7475728155339806
max_depth: 3 : 0.7553398058252427
max_depth: 4 : 0.7533980582524272
max_depth: 5 : 0.7572815533980582


In [ ]:
#hyperparameter tuning for the Random Forest

for estim in range(10, 61, 10):
        model_RF = RandomForestClassifier(n_estimators=estim, random_state=12345)
        model_RF.fit(features_train, target_train)
        score_RF=model_RF.score(features_valid, target_valid)
        print("n_estimators =", estim, ":", score_RF)


n_estimators = 10 : 0.7650485436893204
n_estimators = 20 : 0.7728155339805826
n_estimators = 30 : 0.7766990291262136
n_estimators = 40 : 0.7786407766990291
n_estimators = 50 : 0.7786407766990291
n_estimators = 60 : 0.7786407766990291


In [ ]:
# applying Logistic regression

model_lr=LogisticRegression(random_state=12345)
model_lr.fit(features_train, target_train)
score_lr=model_lr.score(features_valid, target_valid)
print(score_lr)

0.7203883495145631


/opt/conda/lib/python3.7/site-packages/sklearn/linear_model/logistic.py:432: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)


Based on the score, Randomforest with n-estimator=40 looks like the best model with an accuracy of 77.86%. Since n-estimators with 40, 50 and 60 have the same value and as we increase the n-estimators it effect the model processing time, hence I will consider n-estimator=40.

4. Check the quality of the model using the test set.

In [ ]:
model_test=RandomForestClassifier(n_estimators=40, random_state=12345)
model_test.fit(features_train, target_train)
score_test=model_test.score(features_test, target_test)
print(score_test)

0.7884914463452566


The test set has a higher accuracy then the validation set by ~1%